src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -227,28 +390,15 @@ DxcCompilationResult dxcCompile(const CHLSLCompiler* compiler, nbl::asset::hlsl:
 }
 
 
+typedef boost::wave::cpplexer::lex_token<> lex_token_t;
+typedef boost::wave::cpplexer::lex_iterator<lex_token_t> lex_iterator_t;
+typedef boost::wave::context<core::string::iterator, lex_iterator_t, nbl::asset::hlsl::impl::load_file_or_builtin_to_string, nbl::asset::hlsl::impl::custom_preprocessing_hooks> wave_context_t;
+
 std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
 {
-    using lex_token_t = boost::wave::cpplexer::lex_token<>;
-    using lex_iterator_t = boost::wave::cpplexer::lex_iterator<lex_token_t>;
-    using wave_context_t = boost::wave::context<
-        core::string::iterator,
-        lex_iterator_t,
-        boost::wave::iteration_context_policies::load_file_to_string/*,
-        TODO: OurCustomDirectiveHooks -> for pragmas and includes!
-        */
-    >;
 
-    // TODO: change `code` to `const core::string&` because its supposed to be immutable
-    wave_context_t context(code.begin(),code.end(),preprocessOptions.sourceIdentifier.data()/*,TODO: instance of OurCustomDirectiveHooks*/);
-//    context.add_include_path
-//    context.add_sysinclude_path <- for dem builtins! / preprocessOptions.includeFinder?
-//    context.add_macro_definition from preprocessOptions.extraDefines
-    core::string resolvedString;
-    // TODO: fill `resolvedString` with `[context.begin(),context.end()]`
 
-    // Line 1 comes before all the extra defines in the main shader
-    insertIntoStart(code, std::ostringstream(std::string(IShaderCompiler::PREPROC_DIRECTIVE_ENABLER) + "line 1\n"));
+    code.insert(0u, "#line 1\n");
 
     uint32_t defineLeadingLinesMain = 1;
     uint32_t leadingLinesImports = IShaderCompiler::encloseWithinExtraInclGuardsLeadingLines(preprocessOptions.maxSelfInclusionCount + 1u);
```

Comment:
you can remove all this crap, don't add extra defined from `preprocessOptions` this way, add them to the context via the proper boost API